addons/mail/static/src/core/common/composer.js

```diff
@@ -830,12 +834,21 @@ export class Composer extends Component {
     onClickInsertCannedResponse(ev) {
         markEventHandled(ev, "composer.clickInsertCannedResponse");
         const composer = toRaw(this.props.composer);
-        const composerText = composer.composerText;
-        const firstPart = composerText.slice(0, composer.selection.start);
-        const secondPart = composerText.slice(composer.selection.end, composerText.length);
-        const toInsertPart = firstPart.length === 0 || firstPart.at(-1) === " " ? "::" : " ::";
-        composer.composerText = firstPart + toInsertPart + secondPart;
-        this.selection.moveCursor((firstPart + toInsertPart).length);
+        if (this.editor) {
+            if (!isHtmlEmpty(this.props.composer.composerHtml)) {
+                this.editor.shared.dom.insert(" ");
+                this.editor.shared.history.addStep();
```

Comment:
Shouldn't we have a single history step for both operations?
It's a single click = single action, undoing should undo it all?

Reply:
good catch